### ISSUE TITLE
Limit Senate leadership terms

### ISSUE THREAD

**Issue by jcarlson** [mperham/sidekiq#2836]
After a member of a Sidekiq Senate gains leadership of the cluster, that member should not be allowed to retain leadership if 1) that member is stopping (or is quiet) and 2) there are other members of the cluster who are not stopping. If, however, there are no other members of the cluster who are not also stopping, then a coup is unnecessary.

This applies when:

1. Server A is currently leading the senate, comprised of processes A and B 
2. Servers A and B are quieted in preparation for code deploy
3. Servers C and D are added to the cluster with new code base
4. Physical termination of Servers A and B happens at an undetermined point in the future

In this scenario, Servers A and B have been quieted and are expected to shut down; servers C and D are more qualified to become senate leaders, since they are not quieted or stopping.

**Comment by mperham** [comment]
I don't see any easy way to do this.

1. Deploys are signalled via USR1 but they can take many minutes.
2. Cron jobs can be minutely.
3. If a quieted process steps down, it would not create those minutely jobs.
4. This is fine if there's an active process ready to take over but the canonical deploy sees all processes treated the same: all should be quieted and restarted at the same time.

Two Sidekiq processes have no way to tell if one is running a later version of the app codebase, except maybe uptime but even this is not guaranteed (a process supervisor could restart one).

In short, I might be able to make changes to better support your blue/green deploy but I think it would break the semantics other customers are depending on.

**Comment by jcarlson** [comment]
Understood. 

What is the election heartbeat? Aren't minutely cron jobs already a concern if the leader unexpectedly goes away? How long before another instance picks up leadership in a failover scenario?

I don't know if this is the right solution, but what about a mechanism for picking a leader manually? Such as a button in the web UI or a method in the API that allows me to forcefully elect _this_ process instance to be the leader?

I could reasonably write my own Rake tasks that picked the "right" instance to take over, as long as there was some way for me to ask the current leader to step down.

**Comment by jcarlson** [comment]
I looked in the code base a little bit, here's what I found:

- Each process heartbeats to Redis every five seconds, and checks for pending signals
- Senate leaders heartbeat every 15 seconds; followers attempt a coup every 60 seconds
- A process knows if it is quiet/stopping or not

Given this, it should be possible to queue a signal, such as USR3, that each process picks up within 5 seconds, initiating a step down.

Once all members have stepped down, an election should occur naturally within 60 seconds.

At this point, during the election, if a member is "stopping", it should not elect itself. This should not break existing contracts; if all members of the cluster are stopping, no instance would be available to process minutely cron jobs anyway.

Kind of brain dumping here. Does that make sense?

**Comment by mperham** [comment]
Like your database, your job queue is a source of data and will occasionally require a careful, deliberate migration process, just as if you are adding a new column to a table.  It is normal to require two or three deploys to get the complete migration into production.  I'm hesitant to make the deploy process any more complicated than it is today or make changes to that process based on any customer's specific requirements.

Adding more knobs to the leader election is also adding more room to shoot yourself in the foot.  Today it is completely automated and as bulletproof as I can make it.  For example, if you manually assign the leader it becomes a single point of failure.
